File: wrappers/typescript/src/jsonrpc.ts
/**
 * JSON-RPC 2.0 client with per-request-id correlation and notification fanout.
 *
 * Layers JSON-RPC 2.0 semantics on top of any TransportLike (send/onFrame).
 *
 * Design:
 * - call(): allocates a unique request id, creates a Promise, sends the request frame.
 * - dispatch(): routes incoming frames:
 *   - response (has id, no method) → resolve/reject matching pending Promise
 *   - server-initiated request (has id AND method) → call registered handler, send result
 *   - notification (has method, no id) → fanout to all onNotification subscribers
 * - NC-L16 is designed out: each call() has its own independent Promise row in the
 *   pending Map, so two concurrent calls can never interfere.
 */

/** Minimal transport interface: send a frame and register a frame callback. */
export interface TransportLike {
  send(obj: unknown): void;
  onFrame(cb: (obj: unknown) => void): void;
}

/** Notification object fanned out to subscribers. */
export interface Notification {
  method: string;
  params?: unknown;
}

/** Handler for server-initiated requests. Returns the result to send back. */
export type RequestHandler = (params: unknown) => Promise<unknown>;

export class JsonRpcClient {
  private nextId = 1;
  private readonly pending = new Map<
    number,
    { resolve: (value: unknown) => void; reject: (reason: unknown) => void }
  >();
  private readonly notifSubs: Array<(notif: Notification) => void> = [];
  private readonly requestHandlers = new Map<string, RequestHandler>();

  constructor(private readonly transport: TransportLike) {
    transport.onFrame((frame) => this.dispatch(frame));
  }

  /**
   * Send a JSON-RPC 2.0 request and return a Promise that resolves with the result.
   * Allocates a unique id per call — two concurrent calls have independent Promise rows.
   */
  call(method: string, params?: unknown): Promise<unknown> {
    const id = this.nextId++;
    const promise = new Promise<unknown>((resolve, reject) => {
      this.pending.set(id, { resolve, reject });
    });
    this.transport.send({ jsonrpc: "2.0", id, method, params });
    return promise;
  }

  /**
   * Register a subscriber for server-initiated notifications (no id, has method).
   * All subscribers are called for every notification.
   */
  onNotification(cb: (notif: Notification) => void): void {
    this.notifSubs.push(cb);
  }

  /**
   * Register a handler for a specific server-initiated request method.
   * The handler is awaited and its return value is sent back as the result.
   * Unregistered methods receive a -32601 (Method not found) error response.
   */
  onRequest(method: string, handler: RequestHandler): void {
    this.requestHandlers.set(method, handler);
  }

  /** Route an incoming frame to the appropriate handler. */
  private dispatch(frame: unknown): void {
    if (typeof frame !== "object" || frame === null) return;
    const f = frame as Record<string, unknown>;

    const hasId = "id" in f;
    const hasMethod = "method" in f;

    if (hasId && !hasMethod) {
      // Response to a client call (result or error)
      this.handleResponse(f);
    } else if (hasId && hasMethod) {
      // Server-initiated request — dispatch to registered handler
      void this.handleRequest(f);
    } else if (hasMethod && !hasId) {
      // Notification — fanout to subscribers
      this.handleNotification(f);
    }
    // Unknown frame shapes are silently dropped
  }

  private handleResponse(frame: Record<string, unknown>): void {
    const id = frame["id"] as number;
    const pending = this.pending.get(id);
    if (!pending) return;

    this.pending.delete(id);
    if ("error" in frame) {
      pending.reject(frame["error"]);
    } else {
      pending.resolve(frame["result"]);
    }
  }

  private async handleRequest(frame: Record<string, unknown>): Promise<void> {
    const id = frame["id"] as number;
    const method = frame["method"] as string;
    const params = frame["params"];
    const handler = this.requestHandlers.get(method);

    if (!handler) {
      this.transport.send({
        jsonrpc: "2.0",
        id,
        error: { code: -32601, message: "Method not found" },
      });
      return;
    }

    try {
      const result = await handler(params);
      this.transport.send({ jsonrpc: "2.0", id, result });
    } catch (err) {
      this.transport.send({
        jsonrpc: "2.0",
        id,
        error: { code: -32603, message: String(err) },
      });
    }
  }

  private handleNotification(frame: Record<string, unknown>): void {
    const notif: Notification = {
      method: frame["method"] as string,
      params: frame["params"],
    };
    for (const sub of this.notifSubs) {
      sub(notif);
    }
  }
}

File: wrappers/python/src/amplifier_agent_client/jsonrpc.py
"""JSON-RPC 2.0 client with per-request-id correlation and notification fanout.

Layers JSON-RPC 2.0 semantics on top of any _TransportLike (send/on_frame).

Design:
- call(): allocates a unique request id, creates an asyncio.Future, sends the request frame.
- _dispatch(): routes incoming frames:
  - response (has 'id', no 'method') → resolve/reject matching pending Future
  - server-initiated request (has 'id' AND 'method') → call registered handler, send result
  - notification (has 'method', no 'id') → fanout to all on_notification subscribers
- NC-L16 is designed out: each call() has its own independent Future row in _pending,
  so two concurrent calls can never interfere.
"""

from __future__ import annotations

import asyncio
from collections.abc import Callable, Coroutine
from typing import Any, Protocol, runtime_checkable


@runtime_checkable
class _TransportLike(Protocol):
    """Minimal transport interface: send a frame and register a frame callback."""

    def send(self, obj: Any) -> None: ...

    def on_frame(self, cb: Callable[[Any], None]) -> None: ...


# Handler type: async function accepting params, returning the result
RequestHandler = Callable[[Any], Coroutine[Any, Any, Any]]


class JsonRpcClient:
    """JSON-RPC 2.0 client layered on top of any _TransportLike transport."""

    def __init__(self, transport: _TransportLike) -> None:
        self._transport = transport
        self._next_id = 1
        self._pending: dict[int, asyncio.Future[Any]] = {}
        self._notif_subs: list[Callable[[dict[str, Any]], None]] = []
        self._request_handlers: dict[str, RequestHandler] = {}
        # Keep strong references to background request-handling tasks (RUF006).
        self._tasks: set[asyncio.Task[None]] = set()

        transport.on_frame(self._dispatch)

    async def call(self, method: str, params: Any = None) -> Any:
        """Send a JSON-RPC 2.0 request and await the result.

        Allocates a unique id per call — two concurrent calls have independent Futures.
        """
        req_id = self._next_id
        self._next_id += 1

        loop = asyncio.get_running_loop()
        future: asyncio.Future[Any] = loop.create_future()
        self._pending[req_id] = future

        self._transport.send({"jsonrpc": "2.0", "id": req_id, "method": method, "params": params})

        return await future

    def on_notification(self, cb: Callable[[dict[str, Any]], None]) -> None:
        """Register a subscriber for server-initiated notifications.

        All subscribers are called for every notification.
        """
        self._notif_subs.append(cb)

    def on_request(self, method: str, handler: RequestHandler) -> None:
        """Register a handler for a specific server-initiated request method.

        The handler is awaited and its return value is sent back as the result.
        Unregistered methods receive a -32601 (Method not found) error response.
        """
        self._request_handlers[method] = handler

    def _dispatch(self, frame: Any) -> None:
        """Route an incoming frame to the appropriate handler."""
        if not isinstance(frame, dict):
            return

        has_id = "id" in frame
        has_method = "method" in frame

        if has_id and not has_method:
            # Response to a client call (result or error)
            self._handle_response(frame)
        elif has_id and has_method:
            # Server-initiated request — dispatch to registered handler.
            # Keep a strong reference so the task is not garbage-collected (RUF006).
            task = asyncio.create_task(self._handle_request(frame))
            self._tasks.add(task)
            task.add_done_callback(self._tasks.discard)
        elif has_method and not has_id:
            # Notification — fanout to subscribers
            self._handle_notification(frame)
        # Unknown frame shapes are silently dropped

    def _handle_response(self, frame: dict[str, Any]) -> None:
        req_id = frame["id"]
        future = self._pending.pop(req_id, None)
        if future is None or future.done():
            return

        if "error" in frame:
            future.set_exception(Exception(str(frame["error"])))
        else:
            future.set_result(frame.get("result"))

    async def _handle_request(self, frame: dict[str, Any]) -> None:
        req_id = frame["id"]
        method = frame["method"]
        params = frame.get("params")

        handler = self._request_handlers.get(method)
        if handler is None:
            self._transport.send(
                {
                    "jsonrpc": "2.0",
                    "id": req_id,
                    "error": {"code": -32601, "message": "Method not found"},
                }
            )
            return

        try:
            result = await handler(params)
            self._transport.send({"jsonrpc": "2.0", "id": req_id, "result": result})
        except Exception as exc:
            self._transport.send(
                {
                    "jsonrpc": "2.0",
                    "id": req_id,
                    "error": {"code": -32603, "message": str(exc)},
                }
            )

    def _handle_notification(self, frame: dict[str, Any]) -> None:
        notif = {"method": frame["method"], "params": frame.get("params")}
        for sub in self._notif_subs:
            sub(notif)
